hop/configure.py

```diff
@@ -55,22 +61,82 @@ def _add_parser_args(parser):
+    add_cred_parser.add_argument("cred_file", type=str, default=None, nargs='?',
+                                 help="Import credentials from CSV file")
+
+    del_cred_parser = subparser.add_parser("delete-cred", help="Delete a stored credential")
+    del_cred_parser.add_argument("name", type=str,
+                                 help="The name of the credential to delete. This will be treated "
+                                 "as a username, unless it contains a @ character, in which case it"
+                                 " will be split into a username and hostname.")
+
+
+def read_new_credential(csv_file=None):
+    """Import a credential from a CSV file or obtain it interactively from the user.
+
+    Args:
+        csv_file: Path to a file from which to read credential data in CSV format.
+                  If unspecified, the user will be prompted to enter data instead.
+
+    Returns:
+        A configured `auth.Auth` object containing the new credential.
+
+    Raises:
+        FileNotFoundError: If csv_file is not None and refers to a nonexistent path.
+        KeyError: If csv_file is not None and the specified file does not contain either a username
+                  or password field.
+        RuntimeError: If csv_file is None and the interactively entered username or passwod is
+                      empty.
+
+    """
+    if csv_file is None:
+        logger.info("Generating configuration with user-specified username + password")
+        username = input("Username: ")
+        if len(username) == 0:
+            raise RuntimeError("Username may not be empty")
+        password = getpass.getpass()
+        if len(password) == 0:
+            raise RuntimeError("Password may not be empty")
+        hostname = input("Hostname (may be empty): ")
+    else:
+        if os.path.exists(csv_file):
+            with open(csv_file, "r") as f:
+                reader = csv.DictReader(f)
+                cred = next(reader)
+                username = cred["username"]
+                password = cred["password"]
+                hostname = cred["hostname"] if "hostname" in cred else ""
+        else:
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), csv_file)
+    return auth.Auth(username, password, hostname)
+
+
+def write_auth_data(config_file, credentials):
+    """Write configuration file for the set of credentials.
 
-def write_config_file(config_file, username, password):
-    """Write configuration file for the given username and password.
+    Creates containing directories as needed.
 
     Args:
         config_file: configuration file path
-        username: username at hopskotch
-        password: password at hopskotch
+        credentials: list of `auth.Auth` objects representing credentials to be stored
 
     """
+    cred_list = []
+    for cred in credentials:
+        cred_dict = {"username": cred.username, "password": cred._config["sasl.password"]}
```

Comment:
It's definitely not very pretty. We were doing it before in a couple places, I think, but only in tests. 

Alternatives are to duplicate the data in the `Auth` class, which is actually what I previously did with the username, or to make a `password` property which encapsulates this reaching into the guts of the `adc.auth.SASLAuth` inside the `Auth` class. 

Or, I suppose we could ask the upstream ADC Streaming maintainers to add public interfaces for these things, since we do know them.

Reply:
Yeah, that seems a bit tricky. Not sure whether having the password be part of the public API is even a good idea, but I'm not well-versed on the security side. In either case, this is definitely out of scope for this PR so let's punt this for now.